bin/fetchl10nstrings.js

```diff
@@ -0,0 +1,382 @@
+#! /usr/bin/env node
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/**
+ * Create the string bundles and then save them to disk.
+ */
+/*jshint node:true, browser:false, esnext:true*/
+"use strict";
+const async = require("marcosc-async");
+const fs = require("fs");
+const path = require("path");
+const clc = require("cli-color");
+const fetch = require("node-fetch"); // jshint ignore:line
+const repo = "https://hg.mozilla.org/releases/l10n/mozilla-aurora/";
+const newTabPath = "/raw-file/tip/browser/chrome/browser/newTab";
+const globalDirPath = "/raw-file/tip/dom/chrome/global.dtd";
+const l10nPath = path.resolve(`${__dirname}/../l10n/`);
+let defaultLocale;
+
+// CLI-Colors
+const error = clc.red.bold;
+const warn = clc.yellow;
+const notice = clc.blue;
+
+function fetchAndProcessTask(processor) {
+  return async(function*(url) {
+    let response = yield assureResponse(yield fetch(url));
+    let text = yield response.text();
+    return processor(text);
+  });
+}
+
+// Curried functions, reduce duplicate code.
+const dtdProcessorTask = fetchAndProcessTask(processDTD);
+const propProcessorsTask = fetchAndProcessTask(processProps);
+
+/**
+ * Helper object for representing string bundles.
+ *
+ * @constructor
+ * @param {String} locale The locale for the string bundle.
+ */
+function StringBundle(locale) {
+  this.locale = locale;
+}
+
+StringBundle.prototype = {
+  get properties() {
+    return `${repo}${this.locale}${newTabPath}.properties`;
+  },
+  get dtd() {
+    return `${repo}${this.locale}${newTabPath}.dtd`;
+  },
+  get dir() {
+    return `${repo}${this.locale}${globalDirPath}`;
+  },
+  /**
+   * Saves the bundle to disk.
+   *
+   * @return {Promise} Resolves once writing to disk is done is done.
+   */
+  save() {
+    return async.task(function*() {
+      // Kick-off downloads
+      let processedStrings = yield Promise.all([
+        dtdProcessorTask(this.dir),
+        dtdProcessorTask(this.dtd),
+        propProcessorsTask(this.properties)
+      ]);
+      let localeMap = processedStrings.reduce(toSingleMap, new Map());
+      checkForMissingKeys(localeMap, this.locale);
+
+      let defaultClone = new Map(Array.from(defaultLocale.entries()));
+      // Fill in gaps
+      let dirtyMap = toSingleMap(defaultClone, localeMap);
+      let canonicalKeys = Array.from(defaultLocale.keys());
+      let trimmedMap = trimRedundantProps(dirtyMap, canonicalKeys, this.locale);
+      let sortedMap = toSortedMap(trimmedMap);
+      let text = "";
+      try {
+        text = JSON.stringify(sortedMap, mapReplacer, 2);
```

Comment:
Having run this maybe 100 times already and getting the same result (so that's good).  I'm pretty confident we are going to be fine: Changing this behaviour in a JS engine would probably not be web compatible so I doubt it will break any time soon. 

Besides, the ordering is only for our benefit, it doesn't impact the site's behaviour. 

Sent from my iPhone

> On Nov 13, 2015, at 8:31 PM, Olivier Yiptong <notifications@github.com> wrote:
> 
> In bin/fetchl10nstrings.js:
> 
> > +        dtdProcessorTask(this.dir),
> > +        dtdProcessorTask(this.dtd),
> > +        propProcessorsTask(this.properties)
> > +      ]);
> > +      let localeMap = processedStrings.reduce(toSingleMap, new Map());
> > +      checkForMissingKeys(localeMap, this.locale);
> > +
> > +      let defaultClone = new Map(Array.from(defaultLocale.entries()));
> > +      // Fill in gaps
> > +      let dirtyMap = toSingleMap(defaultClone, localeMap);
> > +      let canonicalKeys = Array.from(defaultLocale.keys());
> > +      let trimmedMap = trimRedundantProps(dirtyMap, canonicalKeys, this.locale);
> > +      let sortedMap = toSortedMap(trimmedMap);
> > +      let text = "";
> > +      try {
> > +        text = JSON.stringify(sortedMap, mapReplacer, 2);
> So, apparently, stringify also doesn't guarantee order: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify
> 
> —
> Reply to this email directly or view it on GitHub.
>